File: ml/make-crops-for-labeling.ts
/**
 * Generate per-frame crops for visual ground-truth labeling.
 *
 * For each frame in data/cursor-training-v0/, produce two crops:
 *   1. A 200x200 crop centered on the algorithm's claimed cursor
 *      position. Saved as: _crops/{stem}_crop.jpg
 *   2. A downsampled 2x version of the full frame for context.
 *      Saved as: _crops/{stem}_full.jpg
 *
 * The 200x200 crop is the primary label aid — native pixels, so
 * a 4-5 px cursor is unambiguously visible.
 *
 * Run: npx tsx ml/make-crops-for-labeling.ts
 */
import { promises as fs } from 'fs';
import path from 'path';
import sharp from 'sharp';

const DATA = 'data/cursor-training-v0';
const OUT = path.join(DATA, '_crops');
const CROP_SIZE = 200;
const HALF = CROP_SIZE / 2;

await fs.mkdir(OUT, { recursive: true });

const all = await fs.readdir(DATA);
const jpgs = all.filter((f) => f.endsWith('.jpg')).sort();

let done = 0;
let skipped = 0;
for (const jpg of jpgs) {
  const stem = jpg.replace(/\.jpg$/, '');
  const jsonPath = path.join(DATA, `${stem}.json`);
  const cropPath = path.join(OUT, `${stem}_crop.jpg`);
  const fullPath = path.join(OUT, `${stem}_full.jpg`);

  // Skip if both crops already exist (resumable).
  try {
    await fs.access(cropPath);
    await fs.access(fullPath);
    skipped++;
    continue;
  } catch {
    /* fall through */
  }

  let meta;
  try {
    meta = JSON.parse(await fs.readFile(jsonPath, 'utf8'));
  } catch {
    continue;
  }
  const cx = Math.round(meta.cursor?.x ?? 840);
  const cy = Math.round(meta.cursor?.y ?? 525);

  const img = sharp(path.join(DATA, jpg));
  const { width: W = 1680, height: H = 1050 } = await img.metadata();

  const left = Math.max(0, Math.min(W - CROP_SIZE, cx - HALF));
  const top = Math.max(0, Math.min(H - CROP_SIZE, cy - HALF));

  await sharp(path.join(DATA, jpg))
    .extract({ left, top, width: CROP_SIZE, height: CROP_SIZE })
    .jpeg({ quality: 95 })
    .toFile(cropPath);

  await sharp(path.join(DATA, jpg))
    .resize({ width: Math.floor(W / 2) })
    .jpeg({ quality: 80 })
    .toFile(fullPath);

  // Record crop origin in sidecar so labels can map back.
  meta._crop_origin = { left, top, size: CROP_SIZE };
  await fs.writeFile(jsonPath, JSON.stringify(meta, null, 2));

  done++;
  if (done % 50 === 0) console.error(`  processed ${done} frames`);
}

console.error(`Done: ${done} cropped, ${skipped} already existed.`);

File: ml/make-crops-for-labeling.py
"""
Generate per-frame crops for visual ground-truth labeling.

For each frame in data/cursor-training-v0/, produce two crops:
  1. A 200x200 crop centered on the algorithm's claimed cursor position.
     Saved as: data/cursor-training-v0/_crops/{stem}_crop.jpg
  2. A downsampled 1680x1050 -> 840x525 version of the full frame
     (2x downsample, ~native rendered res). Saved as:
     data/cursor-training-v0/_crops/{stem}_full.jpg

The crop is the primary label aid — at 200x200 native pixels, a
4-5 px cursor is clearly visible (no downsampling).

The full preview lets the human verify "is there ALSO a cursor
elsewhere in the frame that the algorithm missed?"

Run:
  python3 ml/make-crops-for-labeling.py
"""
import json
from pathlib import Path
from PIL import Image

ROOT = Path(__file__).parent.parent
DATA = ROOT / "data" / "cursor-training-v0"
OUT = DATA / "_crops"
OUT.mkdir(exist_ok=True)

CROP_SIZE = 200
HALF = CROP_SIZE // 2

count = 0
for jpg in sorted(DATA.glob("*.jpg")):
    if jpg.parent.name == "_crops":
        continue
    json_path = jpg.with_suffix(".json")
    if not json_path.exists():
        continue
    with open(json_path) as f:
        meta = json.load(f)
    cx = meta["cursor"]["x"]
    cy = meta["cursor"]["y"]
    img = Image.open(jpg)
    W, H = img.size
    left = max(0, min(W - CROP_SIZE, cx - HALF))
    top = max(0, min(H - CROP_SIZE, cy - HALF))
    crop = img.crop((left, top, left + CROP_SIZE, top + CROP_SIZE))
    crop.save(OUT / f"{jpg.stem}_crop.jpg", quality=95)

    full = img.resize((W // 2, H // 2), Image.LANCZOS)
    full.save(OUT / f"{jpg.stem}_full.jpg", quality=85)

    # Record crop origin so labels can map back to frame coords.
    meta["_crop_origin"] = {"left": left, "top": top, "size": CROP_SIZE}
    with open(json_path, "w") as f:
        json.dump(meta, f, indent=2)

    count += 1
    if count % 50 == 0:
        print(f"  processed {count} frames")

print(f"Done: {count} frames cropped → {OUT}")
